File: src/Atlasd/Battlenet/Protocols/BNFTP/BNFTPState.cs
using Atlasd.Battlenet.Protocols.Game;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.BNFTP
{
    class BNFTPState
    {
        public ClientState Client;

        // Version 1
        public UInt16 HeaderLength = 0;
        public UInt16 ProtocolVersion = 0;
        public Platform.PlatformCode PlatformId = Platform.PlatformCode.None;
        public Product.ProductCode ProductId = Product.ProductCode.None;
        public UInt32 AdId = 0;
        public UInt32 AdFileExtension = 0;
        public UInt32 FileStartPosition = 0;
        public UInt64 FileTime = 0;
        public string FileName = null;

        // Version 2
        public UInt32 ServerToken = 0;
        public UInt32 ClientToken = 0;
        public GameKey GameKey = null;

        public BNFTPState(ClientState client)
        {
            Client = client;
            ServerToken = (uint)new Random().Next(0, 0x7FFFFFFF);
        }

        public void Receive(byte[] buffer)
        {
            using var m = new MemoryStream(buffer);
            using var r = new BinaryReader(m);

            /**
             * ## VERSION 1    <-- ##
             * ## Client -> Server ##
             *
             * (TYPE)     (FIELD)                    (DESCRIPTION)
             * UINT16     Request Length
             * UINT16     Protocol Version           0x100 (256) or 0x200 (512)
             */

            HeaderLength = r.ReadUInt16();
            ProtocolVersion = r.ReadUInt16();

            switch (ProtocolVersion)
            {
                case 0x0100:
                    {
                        /**
                         * ## VERSION 1    <-- ##
                         * ## Client -> Server ##
                         *
                         * (TYPE)     (FIELD)                    (DESCRIPTION)
                         * UINT32     Platform ID                See Product Identification
                         * UINT32     Product ID                 See Product Identification
                         * UINT32     Ad Banner ID               0 unless downloading an ad banner
                         * UINT32     Ad Banner File Extension   0 unless downloading an ad banner
                         * UINT32     File start position        For resuming an incomplete download
                         * FILETIME   Filetime
                         * STRING     Filename
                         */

                        PlatformId = (Platform.PlatformCode)r.ReadUInt32();
                        ProductId = (Product.ProductCode)r.ReadUInt32();
                        AdId = r.ReadUInt32();
                        AdFileExtension = r.ReadUInt32();
                        FileStartPosition = r.ReadUInt32();
                        FileTime = r.ReadUInt64();
                        FileName = Encoding.UTF8.GetString(r.ReadByteString());

                        /**
                         * ## VERSION 1    --> ##
                         * ## Server -> Client ##
                         *
                         * (TYPE)     (FIELD)                    (DESCRIPTION)
                         * UINT16     Header Length              Does not include the file length
                         * UINT16     Type
                         * UINT32     File size
                         * UINT32     Ad Banner ID               0 unless downloading an ad banner
                         * UINT32     Ad Banner File Extension   0 unless downloading an ad banner
                         * FILETIME   Filetime
                         * STRING     Filename
                         * VOID       File data
                        */

                        var file = new BNFTP.File(FileName);
                        if (file == null)
                        {
                            Client.Disconnect();
                        }

                        BinaryReader stream = null;

                        bool uploaded = false;
                        try
                        {
                            if (!file.OpenStream())
                            {
                                Client.Disconnect();
                                break;
                            }

                            stream = new BinaryReader(file.StreamReader.BaseStream);
                            stream.BaseStream.Position = Math.Min(stream.BaseStream.Length, FileStartPosition);

                            var fileLength = (int)(stream.BaseStream.Length - stream.BaseStream.Position);
                            HeaderLength = (UInt16)(25 + Encoding.UTF8.GetByteCount(FileName));
                            var outBuf = new byte[HeaderLength];
                            using var wm = new MemoryStream(outBuf);
                            using var w = new BinaryWriter(wm);

                            w.Write((UInt16)HeaderLength);
                            w.Write((UInt16)0); // "Type" ???
                            w.Write((UInt32)fileLength);
                            w.Write((UInt32)AdId);
                            w.Write((UInt32)AdFileExtension);
                            w.Write((UInt64)new FileInfo(file.Path).LastWriteTimeUtc.ToFileTimeUtc());
                            w.Write((string)FileName);

                            Write(outBuf);
                            Write(stream.ReadBytes(fileLength));

                            uploaded = true;
                        }
                        catch (Exception ex)
                        {
                            if (!(ex is IOException || ex is FileNotFoundException || ex is UnauthorizedAccessException || ex is PathTooLongException)) throw;

                            Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_BNFTP, Client.RemoteEndPoint, $"{ex.GetType().Name} error encountered for requested file [{FileName}]" + (string.IsNullOrEmpty(ex.Message) ? "" : $"; message: {ex.Message}"));
                        }
                        finally
                        {
                            if (uploaded)
                            {
                                Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_BNFTP, Client.RemoteEndPoint, $"Uploaded file [{FileName}] ({file.Length} bytes)");
                            }

                            if (stream != null)
                            {
                                stream.Close();
                            }

                            Client.Disconnect();
                        }

                        break;
                    }
                case 0x0200:
                    {
                        /**
                         * ## VERSION 2    <-- ##
                         * ## Client -> Server ##
                         *
                         * (TYPE)     (FIELD)                    (DESCRIPTION)
                         * UINT32     Platform ID                See Product Identification
                         * UINT32     Product ID                 See Product Identification
                         * UINT32     Ad Banner ID               0 unless downloading an ad banner
                         * UINT32     Ad Banner File Extension   0 unless downloading an ad banner
                         * UINT32     File start position        For resuming an incomplete download
                         * FILETIME   Filetime
                         * STRING     Filename
                         */

                        PlatformId = (Platform.PlatformCode)r.ReadUInt32();
                        ProductId = (Product.ProductCode)r.ReadUInt32();
                        AdId = r.ReadUInt32();
                        AdFileExtension = r.ReadUInt32();
                        FileStartPosition = r.ReadUInt32();
                        FileTime = r.ReadUInt64();
                        FileName = Encoding.UTF8.GetString(r.ReadByteString());

                        /**
                         * ## VERSION 2    --> ##
                         * ## Server -> Client ##
                         *
                         * (TYPE)     (FIELD)                    (DESCRIPTION)
                         * UINT32     Server Token
                         */

                        /**
                         * ## VERSION 2    <-- ##
                         * ## Client -> Server ##
                         *
                         * (TYPE)     (FIELD)                    (DESCRIPTION)
                         * UINT32     Starting position          Facilitates resuming
                         * FILETIME   Local filetime
                         * UINT32     Client Token
                         * UINT32     Key Length
                         * UINT32     Key's product value
                         * UINT32     Key's public value
                         * UINT32     Unknown (Always 0)
                         * UINT32 [5] CD key hash
                         * STRING     Filename
                         */

                        /**
                         * ## VERSION 2    --> ##
                         * ## Server -> Client ##
                         *
                         * (TYPE)     (FIELD)                    (DESCRIPTION)
                         * UINT16     Header Length              Does not include the file length
                         * UINT32     File size
                         * UINT32     Ad Banner ID               0 unless downloading an ad banner
                         * UINT32     Ad Banner File Extension   0 unless downloading an ad banner
                         * FILETIME   Filetime
                         * STRING     Filename
                         * VOID       File data
                        */

                        break;
                    }
                default:
                    {
                        Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_BNFTP, $"Received unknown BNFTP protocol version [0x{ProtocolVersion:X4}]");
                        Client.Disconnect("Unknown BNFTP protocol version");
                        break;
                    }
            }
        }

        public void Write(byte[] buffer)
        {
            Client.Send(buffer);
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/BNFTP/BNFTPState.vb
Imports AtlasV.Battlenet.Protocols.Game
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.BNFTP
    Class BNFTPState
        Public Client As ClientState
        Public HeaderLength As UInt16 = 0
        Public ProtocolVersion As UInt16 = 0
        Public PlatformId As Platform.PlatformCode = Platform.PlatformCode.None
        Public ProductId As Product.ProductCode = Product.ProductCode.None
        Public AdId As UInt32 = 0
        Public AdFileExtension As UInt32 = 0
        Public FileStartPosition As UInt32 = 0
        Public FileTime As UInt64 = 0
        Public FileName As String = Nothing
        Public ServerToken As UInt32 = 0
        Public ClientToken As UInt32 = 0
        Public GameKey As GameKey = Nothing

        Public Sub New(ByVal varClient As ClientState)
            Client = varClient
            ServerToken = CUInt(New Random().[Next](0, &H7FFFFFFF))
        End Sub

        Public Sub Receive(ByVal buffer As Byte())
            Using m = New MemoryStream(buffer)
                Using r = New BinaryReader(m)

                    HeaderLength = r.ReadUInt16()
                    ProtocolVersion = r.ReadUInt16()

                    Select Case ProtocolVersion
                        Case &H100
                            PlatformId = CType(r.ReadUInt32(), Platform.PlatformCode)
                            ProductId = CType(r.ReadUInt32(), Product.ProductCode)
                            AdId = r.ReadUInt32()
                            AdFileExtension = r.ReadUInt32()
                            FileStartPosition = r.ReadUInt32()
                            FileTime = r.ReadUInt64()
                            FileName = Encoding.UTF8.GetString(r.ReadByteString())
                            Dim file = New BNFTP.File(FileName)

                            If file Is Nothing Then
                                Client.Disconnect()
                            End If

                            Dim stream As BinaryReader = Nothing
                            Dim uploaded As Boolean = False

                            Try

                                If Not file.OpenStream() Then
                                    Client.Disconnect()
                                    Exit Select
                                End If

                                stream = New BinaryReader(file.StreamReader.BaseStream)
                                stream.BaseStream.Position = Math.Min(stream.BaseStream.Length, FileStartPosition)
                                Dim fileLength = CInt((stream.BaseStream.Length - stream.BaseStream.Position))
                                HeaderLength = CUShort((25 + Encoding.UTF8.GetByteCount(FileName)))
                                Dim outBuf = New Byte(HeaderLength - 1) {}

                                Using wm = New MemoryStream(outBuf)
                                    Using w = New BinaryWriter(wm)
                                        w.Write(CUShort(HeaderLength))
                                        w.Write(CUShort(0))
                                        w.Write(CUInt(fileLength))
                                        w.Write(CUInt(AdId))
                                        w.Write(CUInt(AdFileExtension))
                                        w.Write(CULng(New FileInfo(file.Path).LastWriteTimeUtc.ToFileTimeUtc()))
                                        w.Write(CStr(FileName))
                                    End Using
                                End Using

                                Write(outBuf)
                                Write(stream.ReadBytes(fileLength))
                                uploaded = True
                            Catch ex As Exception
                                If Not (TypeOf ex Is IOException OrElse TypeOf ex Is FileNotFoundException OrElse TypeOf ex Is UnauthorizedAccessException OrElse TypeOf ex Is PathTooLongException) Then Throw
                                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_BNFTP, Client.RemoteEndPoint, $"{ex.[GetType]().Name} error encountered for requested file [{FileName}]" & (If(String.IsNullOrEmpty(ex.Message), "", $"; message: {ex.Message}")))
                            Finally

                                If uploaded Then
                                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_BNFTP, Client.RemoteEndPoint, $"Uploaded file [{FileName}]")
                                End If

                                If stream IsNot Nothing Then
                                    stream.Close()
                                End If

                                Client.Disconnect()
                            End Try

                            Exit Select
                        Case &H200
                            PlatformId = CType(r.ReadUInt32(), Platform.PlatformCode)
                            ProductId = CType(r.ReadUInt32(), Product.ProductCode)
                            AdId = r.ReadUInt32()
                            AdFileExtension = r.ReadUInt32()
                            FileStartPosition = r.ReadUInt32()
                            FileTime = r.ReadUInt64()
                            FileName = Encoding.UTF8.GetString(r.ReadByteString())
                            Exit Select
                        Case Else
                            Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_BNFTP, $"Received unknown BNFTP protocol version [0x{ProtocolVersion}]")
                            Client.Disconnect("Unknown BNFTP protocol version")
                            Exit Select
                    End Select

                End Using
            End Using
        End Sub

        Public Sub Write(ByVal buffer As Byte())
            Client.Send(buffer)
        End Sub
    End Class
End Namespace